clients/src/main/java/org/apache/kafka/common/InvalidRecordException.java

```diff
@@ -18,16 +18,40 @@
 
 import org.apache.kafka.common.errors.ApiException;
 
+import java.util.Collections;
+import java.util.Map;
+
 public class InvalidRecordException extends ApiException {
 
     private static final long serialVersionUID = 1;
 
+    // Relative offset of the record that throws this exception
+    private Map<Integer, String> errorRecords = Collections.emptyMap();
+
     public InvalidRecordException(String s) {
         super(s);
     }
 
+    public InvalidRecordException(String s, Map<Integer, String> errorRecords) {
+        super(s);
+        this.errorRecords = errorRecords;
+    }
+
     public InvalidRecordException(String message, Throwable cause) {
         super(message, cause);
     }
 
+    public InvalidRecordException(String message, Throwable cause, Map<Integer, String> errorRecords) {
+        super(message, cause);
+        this.errorRecords = errorRecords;
+    }
+
+    public void setErrorRecords(Map<Integer, String> errorRecords) {
+        this.errorRecords = errorRecords;
+    }
+
+    public Map<Integer, String> getErrorRecords() {
```

Comment:
nit: kafka apis typically drop the `get` so this would just be `errorRecords`.

By the way, if we are anticipating that eventually we may have error codes, then perhaps we should wrap the error message in an object. For example,
```java
RecordError {
  private final String message;
  private final int index;
}
```
Then this API can be the following:
```
public List<RecordError> getRecordErrors();
```
Similarly for `InvalidTimestampException`.

Reply:
Thinking about this a bit more, this API is mostly for internal use, right? A user can't do anything with the batch indexes. So should we avoid letting this leak into the API?